resteasy-core/src/main/java/org/jboss/resteasy/plugins/providers/ProviderHelper.java

```diff
@@ -128,4 +133,55 @@ public static void writeTo(final InputStream in, final OutputStream out) throws
          out.write(buf, 0, read);
       }
    }
+
+   /**
+    * @param in input stream
+    * @param out output stream
+    * @throws IOException if I/O error occurred
+    */
+   public static CompletionStage<Void> writeToAndCloseInput(final InputStream in, final AsyncOutputStream out)
+   {
+      return writeTo(in, out).whenComplete((v, t) -> {
+         try {
+            in.close();
+         } catch(IOException x) {
+            throw new RuntimeException(x);
+         }
+      });
+   }
+
+   /**
+    * @param in input stream
+    * @param out output stream
+    * @throws IOException if I/O error occurred
+    */
+   public static CompletionStage<Void> writeTo(final InputStream in, final AsyncOutputStream out)
+   {
+      final byte[] buf = new byte[2048];
+      return writeTo(in, out, buf);
+   }
+
+   private static CompletionStage<Void> writeTo(final InputStream in, final AsyncOutputStream out, byte[] buf)
+   {
+      int read;
+      try
+      {
+         if ((read = in.read(buf)) != -1)

```

Comment:
InputStream.read is a blocking operation, so if this is called from the IO thread it could block and cause performance problems. It probably won't cause deadlocks, unless the input stream is also backed by the same IO thread (e.g. the Undertow input stream).

The only way to really fix this would be to move this work to an executor if you are on the IO thread.

Reply:
So yeah, I know input is broken too because we need async IO support. I plan to do this later though as it wasn't as urgent and this has taken enough time.

It's entirely possible (and very likely) that the input in question is not backed by an IO thread, in which case yeah we're blocking the IO thread, but not deadlocking it.

If it is backed by an IO thread, I don't have the possibility to detect it ATM. I would prefer to delay solving this until later.

Reply:
I agree that we can ignore this for now. If anyone actually hits it we can tell them not to return a blocking stream from an IO thread, but I don't think there is a real use case where you would run into this.